### PR TITLE
Chapter 7 Carlos corrections

### PR DESCRIPTION
**Pull request by Pattrigue**

### CODE REVIEW

File: report/sections/TestRunner/chapter.tex [hunk 1]
```diff
@@ -1,33 +1,32 @@
...
 The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
 Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.
 
-Sequentially starting a new \texttt{runhaskell} process for each request was identified as a major bottleneck during testing, which will be described in detail in chapter \ref{chap:Benchmarking}.
-In order to address this, we started by refactoring the code base to allow running several instances of the GHC interpreter process at the same time.
-However, despite this running faster for a small number of clients, this approach in itself is not sufficiently stable:
-If multiple clients were to post a code submission at the same time, the Test Runner will start a new process for each request.
+Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
+Processing each request sequentially results in slow response times, which may also lead to connection timeouts.
+In order to address this,
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:16]: unfinished sentence

File: report/sections/TestRunner/chapter.tex [hunk 2]
```diff
@@ -1,33 +1,32 @@
...
-
-The Test Runner exposes an endpoint \texttt{/haskell/submit/}, which is used to post code submissions.
-Originally, we used the Rust crate Rocket to implement this endpoint.
-However, we eventually switched to using the axum crate instead, as it is developed and maintained by the same developers behind the Tokio package we make heavy use of for asynchronous programming and multithreading.
-Furthermore, it was deemed easier to implement a shared state among endpoints by leveraging axum's \texttt{layer} feature.
-We give an overview and a more technical explanation of how this was done in section \ref{sec:queue-system}.
-
-When the Test Runner receives the request from Next.js, its singular goal is to run the tests against the code and send the result of this back as fast as possible.
-In order to run the tests, we need to run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
+When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and send back the result as fast as possible.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:2]:
```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and make the result available as fast as possible.
```
We don't send the results back without pollling, and that violates the goal criteria otherwise.

File: report/sections/TestRunner/chapter.tex [hunk 3]
```diff
@@ -1,33 +1,32 @@
...
+In order to run the tests, we run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:3]:
```suggestion
In order to run the tests, we run the GHC interpreter on the test code, which, in turns, needs access to the exercise submission code.
```

File: report/sections/TestRunner/chapter.tex [hunk 4]
```diff
@@ -1,33 +1,32 @@
...
 The GHC interpreter can be run from the command line with arguments specifying the path Haskell file to run.
-In other words, the Test Runner needs to write the user submission and corresponding test code to a file.
-However, if more users were to make a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
-
+In other words, the Test Runner writes the user submission and corresponding test code to a file.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:5]:
```suggestion
So to run the tests, the Test Runner writes the user's code and corresponding exercise test code to a file.
```
This feels clearer. Removing `needs` from the sentence changed its meaning, which IMO made it a bit wonky.

And should we clarify that it's actually writing to separate files?

File: report/sections/TestRunner/chapter.tex [hunk 5]
```diff
@@ -1,33 +1,32 @@
...
 This way, we can pass the path to this isolated directory to the GHC interpreter process as an argument.
 Once the tests have been run, the directory and its contents are deleted again.
 The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
 Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.
 
-Sequentially starting a new \texttt{runhaskell} process for each request was identified as a major bottleneck during testing, which will be described in detail in chapter \ref{chap:Benchmarking}.
-In order to address this, we started by refactoring the code base to allow running several instances of the GHC interpreter process at the same time.
-However, despite this running faster for a small number of clients, this approach in itself is not sufficiently stable:
-If multiple clients were to post a code submission at the same time, the Test Runner will start a new process for each request.
+Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:14]:
```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when users submit solutions in quick succession.
```

File: report/sections/TestRunner/chapter.tex [hunk 6]
```diff
@@ -1,33 +1,32 @@
...
+Processing each request sequentially results in slow response times, which may also lead to connection timeouts.
+In order to address this,
+This was solved by allowing multiple instances of the GCH interpreter process to run at the same time.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:17]:
```suggestion
We solved the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```
Clarifying which problem was solved.

File: report/sections/TestRunner/chapter.tex [hunk 7]
```diff
@@ -1,33 +1,32 @@
...
+However, despite this being faster for a small number of clients, this approach in itself is not sufficient:
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:18]:
```suggestion
However, despite this being faster for a small number of clients, this approach in itself is not sufficient.
```

File: report/sections/TestRunner/chapter.tex [hunk 8]
```diff
@@ -1,33 +1,32 @@
...
+If multiple clients post a code submission at the same time, the Test Runner will start a new process for each request.
 Given a large enough number of requests, this would crash the Test Runner web server.
 Therefore, we needed a way to limit the number of running GHC interpreter processes while at the same time storing subsequent requests for future processing.
 
+To solve the aforementioned problem, we decided to implement a queue system.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:23]:
```suggestion
To prevent crashes, we decided to implement a queue system.
```
In just a few lines, we mention two problems and make references to solutions. Might be better to specify which problem we're solving, like so?

**Review by MrHonket — DISMISSED** [review]
1 very minor change, otherwise it looks great

File: report/sections/TestRunner/chapter.tex [hunk 9]
```diff
@@ -1,33 +1,32 @@
...
+The idea is to enqueue requests for future processing if the number of currently running GHC interpreter processes reaches an upper limit.
+This limit is defined as an environmental constant in the program.
+However, this introduces a new problem.
+The server can no longer directly respond to a POST request with the Test Runner results because the request may be handled at a later point in time.
+This means that if the request were enqueued for long enough, it would eventually time out.
```
Comment (chhoumann) [inline report/sections/TestRunner/chapter.tex:28]:
```suggestion
However, this introduced a new problem.
The server could no longer respond directly to a POST request with the Test Runner results because the request may be handled at a later point in time.
This means that if the request were enqueued for long enough, it would eventually time out.
```
tense

File: report/sections/TestRunner/queue-system.tex [hunk 10]
```diff
@@ -1,5 +1,8 @@
 \section{Queue System} \label{sec:queue-system}
-We have two endpoints that need to share data, and the Rust programming language does not allow global variables.
+The idea with the queue system is that a specified number of concurrent threads in a thread pool continuously receive and process code submissions.
+Each submission is then processed by one of these threads, and the result of the Test Runner is packaged into a hashmap where the UUID serves as a key to look up the result.
```
Comment (chhoumann) [inline report/sections/TestRunner/queue-system.tex:3]:
```suggestion
Each submission is then processed by one of these threads, and the result of the Test Runner is packaged into a hashmap where the UUID token serves as a key to look up the result.
```
For clarity

**Review by chhoumann — CHANGES_REQUESTED** [review]
Great changes. Made the flow of this chapter better, and the main points clearer. I have some suggestions that might add clarity.

File: report/sections/TestRunner/chapter.tex (hunk 8 above)
Comment (TheRumle) [reply report/sections/TestRunner/chapter.tex:23]: Maaske udvide saetningen. "implement a queue system enabling xxxx"

File: report/sections/TestRunner/chapter.tex (hunk 5 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:14]: The idea with "multiple" was to say that it was many users at the same time. Maybe we should just change "multiple" to many instead of removing it?

File: report/sections/TestRunner/chapter.tex (hunk 1 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex]: Removed it.

File: report/sections/TestRunner/chapter.tex [hunk 11]
```diff
@@ -1,33 +1,31 @@
...
 The \texttt{stdout} of the process is then read to determine whether the tests succeeded or not.
 Finally, the interpreter output and the test success flag is packaged into a JSON object and sent back to the client.
 
-Sequentially starting a new \texttt{runhaskell} process for each request was identified as a major bottleneck during testing, which will be described in detail in chapter \ref{chap:Benchmarking}.
-In order to address this, we started by refactoring the code base to allow running several instances of the GHC interpreter process at the same time.
-However, despite this running faster for a small number of clients, this approach in itself is not sufficiently stable:
-If multiple clients were to post a code submission at the same time, the Test Runner will start a new process for each request.
+Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when multiple users submit solutions in quick succession.
+Processing each request sequentially results in slow response times, which may also lead to connection timeouts.
+This was solved by allowing multiple instances of the GCH interpreter process to run at the same time.
```
Comment (Pattrigue) [inline report/sections/TestRunner/chapter.tex:16]:
```suggestion
We solved this the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```

File: report/sections/TestRunner/chapter.tex (hunk 6 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:16]:
```suggestion
We solved this the timeout problem by allowing multiple instances of the GHC interpreter process to run at the same time.
```

File: report/sections/TestRunner/chapter.tex (hunk 8 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:22]:
```suggestion
To address the aforementioned problem where the system crashes under heavy use, we decided to implement a queue system.
```
Clarify that the system doesn't "just crash", but only does so under heavy use.

File: report/sections/TestRunner/chapter.tex [hunk 12]
```diff
@@ -1,33 +1,31 @@
...
-
-The Test Runner exposes an endpoint \texttt{/haskell/submit/}, which is used to post code submissions.
-Originally, we used the Rust crate Rocket to implement this endpoint.
-However, we eventually switched to using the axum crate instead, as it is developed and maintained by the same developers behind the Tokio package we make heavy use of for asynchronous programming and multithreading.
-Furthermore, it was deemed easier to implement a shared state among endpoints by leveraging axum's \texttt{layer} feature.
-We give an overview and a more technical explanation of how this was done in section \ref{sec:queue-system}.
-
-When the Test Runner receives the request from Next.js, its singular goal is to run the tests against the code and send the result of this back as fast as possible.
-In order to run the tests, we need to run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
+When the Test Runner receives the request from the backend component, its singular goal is to run the tests against the code and send back the result as fast as possible.
```
Comment (Pattrigue) [inline report/sections/TestRunner/chapter.tex:2]:
```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests on the submitted code and make the result available as fast as possible.
```

File: report/sections/TestRunner/chapter.tex (hunk 4 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:5]:
```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to a file.
```

File: report/sections/TestRunner/chapter.tex (hunk 2 above)
Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:2]:
```suggestion
When the Test Runner receives the request from the backend component, its singular goal is to run the tests on the submitted code and make the result available as fast as possible.
```

File: report/sections/TestRunner/chapter.tex (hunk 5 above)
Comment (chhoumann) [reply report/sections/TestRunner/chapter.tex:14]: So it is to emphasize that it is _many_ users, not just some number of users? If so, sure, we can change multiple to another word.

Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:14]:
```suggestion
Sequentially starting a new \texttt{runhaskell} process for each request is a major bottleneck when many multiple users submit solutions in quick succession.
```

File: report/sections/TestRunner/chapter.tex (hunk 4 above)
Comment (chhoumann) [reply report/sections/TestRunner/chapter.tex:5]:
```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to files.
```
I don't think `a file each` is great. What do you think of this formulation?
or `separate files`?

Comment (Pattrigue) [reply report/sections/TestRunner/chapter.tex:5]:
```suggestion
So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to separate files.
```

**Review by chhoumann — APPROVED** [review]

File: report/sections/TestRunner/chapter.tex [hunk 13]
```diff
@@ -1,33 +1,31 @@
...
-
-When the Test Runner receives the request from Next.js, its singular goal is to run the tests against the code and send the result of this back as fast as possible.
-In order to run the tests, we need to run the GHC interpreter on the test code, which, in turns, need access to the exercise submission code.
+When the Test Runner receives the request from the backend component, its singular goal is to run the tests on the submitted code and make the result available as fast as possible.
+In order to run the tests, we run the GHC interpreter on the test code, which, in turns, needs access to the exercise submission code.
 The GHC interpreter can be run from the command line with arguments specifying the path Haskell file to run.
-In other words, the Test Runner needs to write the user submission and corresponding test code to a file.
-However, if more users were to make a submission at the same time, the content of the file could potentially be overwritten while the GHC interpreter is running the tests.
-
+So to run the tests, the Test Runner writes the code submitted by the user, and the accompanying exercise test code, to separate files.
```
Comment (Pattrigue) [inline report/sections/TestRunner/chapter.tex:5]:
```suggestion
So to run the tests, the Test Runner writes the code submitted by the user and the accompanying test code to separate files.
```
